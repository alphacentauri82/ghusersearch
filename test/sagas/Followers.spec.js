import {
  getFollowers,
} from 'store/sagas/Followers';
import {
  put,
} from 'redux-saga/effects';

describe('Saga: getFollowers', () => {

  describe('when no cached data exists', () => {
    it('should call the API with the url and parameters', () => {
      const generator = getFollowers({payload: {url: 'followers.net'}});
      generator.next();
      const followers = generator.next().value;
      expect(followers.CALL).toMatchSnapshot();
    });

    it('should put a success action with the normalized data', () => {
      const generator = getFollowers({payload: {url: 'followers.net'}});
      const response = {
        data: [
          {id: 1, login: 'test'},
          {id: 2, login: 'hello'},
        ],
      };
      generator.next();
      generator.next();
      const normalized = generator.next(response).value;
      expect(normalized.PUT.action).toMatchSnapshot();
    });
  });

  describe('when cached data does exist', () => {
    it('should pass the result to the success action and not call the API', () => {
      const generator = getFollowers({payload: {url: 'followers.net'}});
      generator.next();
      const successAction = generator.next(
        {
          result: [1, 2],
          entities: {},
        }
      ).value;

      expect(successAction).toEqual(
        put({
          meta: {fromCache: true},
          payload: {
            entities: {},
            result: [1, 2],
            url: 'followers.net',
          },
          type: 'FOLLOWERS_SUCCESS',
        })
      );
    });
  });

  describe('when a request to the API fails', () => {
    it('should call a failure action', () => {
      const generator = getFollowers({payload: {url: 'followers.net'}});
      const apiError = new Error('it went wrong');
      apiError.response = 'test';

      generator.next();
      generator.next();
      const errorAction = generator.throw(apiError).value;
      expect(errorAction.PUT).toMatchSnapshot();
    });
  });

});
