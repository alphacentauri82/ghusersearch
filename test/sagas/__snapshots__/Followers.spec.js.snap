// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga: getFollowers when a request to the API fails should call a failure action 1`] = `
Object {
  "action": Object {
    "error": true,
    "payload": [Error: it went wrong],
    "type": "FOLLOWERS_FAILURE",
  },
  "channel": null,
}
`;

exports[`Saga: getFollowers when no cached data exists should call the API with the url and parameters 1`] = `
Object {
  "args": Array [
    "followers.net",
    Object {
      "params": Object {
        "per_page": 8,
      },
    },
  ],
  "context": null,
  "fn": [Function],
}
`;

exports[`Saga: getFollowers when no cached data exists should put a success action with the normalized data 1`] = `
Object {
  "meta": Object {
    "fromCache": false,
  },
  "payload": Object {
    "entities": Object {
      "users": Object {
        "1": Object {
          "id": 1,
          "login": "test",
        },
        "2": Object {
          "id": 2,
          "login": "hello",
        },
      },
    },
    "result": Array [
      1,
      2,
    ],
    "url": "followers.net",
  },
  "type": "FOLLOWERS_SUCCESS",
}
`;
